### PR TITLE
Add IntegerKey transaction family specification

### PR DESCRIPTION
**Pull request by dplumb94**
Signed-off-by: Darian Plumb <dplumb@bitwise.io>

### CODE REVIEW

File: docs/source/transaction_family_specifications/integerkey_transaction_family.rst [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+* Address of the *Name* being changed or added
+
+
+The outputs for IntegerKey family transactions must include:
+
+* Address of the *Name* being changed or added
+
+Dependencies
+------------
+* A list of transaction *header_signatures* which are required dependencies and must be processed prior to processing this transaction
```
Comment (jsmitchell) [inline docs/source/transaction_family_specifications/integerkey_transaction_family.rst]: You might want to add "For example, 'inc' and 'dec' transactions might want to list the initial 'set' transaction for the *Name*. If 'inc' or 'dec' transactions are ordered before the corresponding 'set' transaction (and don't list the 'set' transaction as a dependency, they will be considered invalid (because *Name* will not exist when they are processed)."

File: docs/source/transaction_family_specifications/integerkey_transaction_family.rst [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+
+Execution
+=========
+
+The IntegerKey transaction processor receives a transaction request and a state.
+If the payload of the transaction request is empty, the transaction is invalid.
+
+The address for the transaction is generated using the algorithm stated in the Addressing 
+section of this document. If an encoding error occurs, it is considered an internal 
+error.
```
Comment (jsmitchell) [inline docs/source/transaction_family_specifications/integerkey_transaction_family.rst]: "it is considered an internal error, and the transaction is invalid."

File: docs/source/transaction_family_specifications/integerkey_transaction_family.rst [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+    })
+
+\ *Names* and *Values* must conform to the following rules:
+
+* Valid *Names* are utf-8 encoded strings with a maximum of 20 characters
+* Valid *Values* must be integers in the range of 0 through 2\ :sup:`32` − 1 (32-bit unsigned int)
+
+Addressing
+----------
+When an IntegerKey family transaction entry is stored, it can be accessed by addressing using
```
Comment (jsmitchell) [inline docs/source/transaction_family_specifications/integerkey_transaction_family.rst]: Maybe reword this to "IntegerKey data is stored in state using addresses which are generated from the intkey namespace prefix and the unique *Name* of the intkey variable. These addresses are generated using the following algorithm:"

File: docs/source/transaction_family_specifications/integerkey_transaction_family.rst [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+When an IntegerKey family transaction entry is stored, it can be accessed by addressing using 
+the following algorithm:
+
+The address of an IntegerKey family transaction entry must be a 70 character hexadecimal string. 
+Addresses for the IntegerKey transaction family are generated by adding the last 64 characters 
+of the sha512 hash of the entry *Name* to the first 6 characters of a sha512 hash of the 
+IntegerKey namespace (the IntegerKey namespace is the string "intkey"). For example, an 
+address could be generated like this:
+
+.. code-block:: pycon
```
Comment (ojalatodd) [inline docs/source/transaction_family_specifications/integerkey_transaction_family.rst]: typo: pycon ==> python

File: docs/source/transaction_family_specifications/integerkey_transaction_family.rst (hunk 1 above)
Comment (ojalatodd) [inline docs/source/transaction_family_specifications/integerkey_transaction_family.rst]: Agree with @jsmitchell  on  a little example here.

File: docs/source/transaction_family_specifications/integerkey_transaction_family.rst (hunk 2 above)
Comment (boydjohnson) [inline docs/source/transaction_family_specifications/integerkey_transaction_family.rst]: @jsmitchell, So with InternalErrors, they are currently being treated as if the txn is invalid, but according to @vaporos that is not the desired intent. After this coming sprint they won't be the same as an invalid transaction. They should be used in situations where the txn handler can't determine the validity of the txn. @agunde406  has a story in the upcoming sprint to change the behavior on the validator side to keep trying indefinitely (we may have to come up with a better solution than this) the transaction with a transaction processor if it comes back with InternalError. We should be pretty careful where we use InternalError.

Comment (jsmitchell) [inline docs/source/transaction_family_specifications/integerkey_transaction_family.rst]: In that case, it sounds like we should strike any language in this doc referring to InternalError, and just say "the transaction is invalid".

File: docs/source/transaction_family_specifications/integerkey_transaction_family.rst [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+
+    >>> hashlib.sha512('intkey'.encode('utf-8')).hexdigest()[0:6] + hashlib.sha512('eBuPof'.encode('utf-8')).hexdigest()[:64]
```
Comment (boydjohnson) [inline docs/source/transaction_family_specifications/integerkey_transaction_family.rst]: Up above it says 'last 64 characters of the sha512 hash...', but this code shows the __first__ 64 characters of the (hex digest of the) sha512 hash being taken. '[:64]' is the first 64 characters, while '[-64:]' would be the last 64 characters.

File: docs/source/transaction_family_specifications/integerkey_transaction_family.rst [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+The 'set' verb is used to create new entries. The initial value of the entry will be 
+set to the value specified in the transaction request. The 'inc' and 'dec' verbs are 
+used to change the value of existing entries in state.
+
+State
+=====
+This section describes in detail how IntegerKey transaction information is stored 
+and addressed.
+
+The address values of IntegerKey transaction family entries are stored in state as
```
Comment (boydjohnson) [inline docs/source/transaction_family_specifications/integerkey_transaction_family.rst]: I would mention, maybe as a note or an aside, that there could be multiple Names in the state dictionary for a particular address. This allows for the possibility of hash collisions which may be *possible*. Or we could do away with this requirement that we allow for the possibility of hash collisions, and mention that in the spec. Either way it should be mentioned in the spec.

File: docs/source/transaction_family_specifications/integerkey_transaction_family.rst [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+a cbor encoded dictionary, with the key being the *Name* and the value being an integer 
+*Value*.
+
+.. code-block:: python
+
+    cbor({
```
Comment (boydjohnson) [inline docs/source/transaction_family_specifications/integerkey_transaction_family.rst]: So the cbor specification is making more sense to me. The first type of map is described this way: ```0xa0..0xb7      | map (0x00..0x17 pairs of data items follow) ```. So that is the first hex representation of the initial bytes for a map that has 23 or less key-value pairs in a map. So when the maps have 23 or less pairs, this would specify the format of the bytes stored in state. The other map specifications in cbor have to do with more pairs of items, and then also an indefinite amount.

Comment (boydjohnson) [inline docs/source/transaction_family_specifications/integerkey_transaction_family.rst]: ^ wasn't a good comment, but was more information. I would reference the cbor specification at https://tools.ietf.org/html/rfc7049 and say that we are using a definite map, with that snippet of code up above, and maybe give examples of possible bytes stored in state. So using a python3 repl, b'\xa1cfoo\x13' = cbor.dumps({'foo': 13})

**Comment by jsmitchell** [comment]
@dplumb94 note that you are getting a docs build error on this.

```
! Package inputenc Error: Unicode char − (U+2212)
(inputenc)                not set up for use with LaTeX.

See the inputenc package documentation for explanation.
Type  H <return>  for immediate help.
 ...                                              
                                                  
l.6364 ...e range of 0 through 2$^{\text{32}}$ −
                                                   1 (32-bit unsigned int)
? 
! Emergency stop.
 ...                                              
                                                  
l.6364 ...e range of 0 through 2$^{\text{32}}$ −
                                                   1 (32-bit unsigned int)
!  ==> Fatal error occurred, no output PDF file produced!```

You can test this locally in your vagrant vm by making sure that `sudo tools/plugins/install_latex.sh` has been run and then doing a `make latexpdf` in sawtooth-core/docs/

File: docs/source/transaction_family_specifications/integerkey_transaction_family.rst (hunk 3 above)
Comment (peterschwarz) [inline docs/source/transaction_family_specifications/integerkey_transaction_family.rst]: Looks like the build failure is due to "2\ :sup:`32`"

File: docs/source/transaction_family_specifications/integerkey_transaction_family.rst [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+If the *Verb* is 'inc', the *Name* specified by the transaction request is checked determine 
+if the entry exists in state. If the *Name* does not exist in state, it is an invalid transaction. 
+Otherwise, we attempt to increment the *Value* in state by the *Value* specified in the transaction 
+request. If this incrementation would result in a value outside the range of 0 through 2\ :sup:`32` − 1 
+it is considered an invalid transaction. Otherwise, the *Value* in state is incremented.
+
+If the *Verb* is 'dec', the *Name* specified by the transaction request is checked determine 
+if the entry exists in state. If the *Name* does not exist in state, it is an invalid transaction. 
+Otherwise, we attempt to decrement the *Value* in state by the *Value* specified in the transaction 
+request. If this decrementation would result in a value outside the range of 0 through 2\ :sup:`32` − 1,
```
Comment (peterschwarz) [inline docs/source/transaction_family_specifications/integerkey_transaction_family.rst]: Another instance of "2\ :sup:`32`" that is could be causing an issue

File: docs/source/transaction_family_specifications/integerkey_transaction_family.rst [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+a 32-bit unsigned integer. If any of these checks fail, the transaction is invalid.
+
+If the *Verb* is 'set', the state is checked to determine if the *Name* associated with the 
+transaction request already exists. If it does already exist, the transaction is invalid. 
+Otherwise the *Name* and *Value* are stored as a new entry in state.
+
+If the *Verb* is 'inc', the *Name* specified by the transaction request is checked determine 
+if the entry exists in state. If the *Name* does not exist in state, it is an invalid transaction. 
+Otherwise, we attempt to increment the *Value* in state by the *Value* specified in the transaction 
+request. If this incrementation would result in a value outside the range of 0 through 2\ :sup:`32` − 1
```
Comment (peterschwarz) [inline docs/source/transaction_family_specifications/integerkey_transaction_family.rst]: Another instance of "2\ :sup:`32`" that is could be causing an issue

File: docs/source/transaction_family_specifications/integerkey_transaction_family.rst [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+set to the value specified in the transaction request. The 'inc' and 'dec' verbs are 
+used to change the value of existing entries in the state dictionary.
+
+State
+=====
+This section describes in detail how IntegerKey transaction information is stored 
+and addressed.
+
+The address values of IntegerKey transaction family entries are stored in state as
+a cbor encoded dictionary, with the key being the *Name* and the value being an integer
```
Comment (peterschwarz) [inline docs/source/transaction_family_specifications/integerkey_transaction_family.rst]: In text, outside of the example, cbor should be all caps: CBOR

File: docs/source/transaction_family_specifications/integerkey_transaction_family.rst [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+* Valid *Names* are utf-8 encoded strings with a maximum of 20 characters
+* Valid *Values* must be integers in the range of 0 through 2\ :sup:`32` - 1 (32-bit unsigned int)
+
+Addressing
+----------
+IntegerKey data is stored in the state dictionary using addresses which are generated from the IntegerKey namespace prefix and the unique name of the IntegerKey entry. These addresses are generated using the following algorithm:
+
+.. code-block:: pycon
+
+    >>> hashlib.sha512('intkey'.encode('utf-8')).hexdigest()[0:6] + hashlib.sha512('name'.encode('utf-8')).hexdigest()[-64:]
```
Comment (nick-drozd) [inline docs/source/transaction_family_specifications/integerkey_transaction_family.rst]: Should that `[-64:]` be `[64:]`? `hashlib.sha512('name'.encode('utf-8')).hexdigest()` is 128 characters, so it works out the same, but I'm not sure which is intended. It might be helpful to include an English description of the algorithm.

File: docs/source/transaction_family_specifications/integerkey_transaction_family.rst [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+
+Transaction Payload
+===================
+
+IntegerKey transaction request payloads are defined by the following CBOR data format:
+
+.. code-block:: python
+
+    cbor.dumps({
+        // Describes the action the transaction takes, either 'set', 'inc', or 'dec'
```
Comment (nick-drozd) [inline docs/source/transaction_family_specifications/integerkey_transaction_family.rst]: Python comments use `#`, not `//`.

File: docs/source/transaction_family_specifications/integerkey_transaction_family.rst [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+
+    cbor.dumps({
+        // The name of the entry : The current value of the entry
+        'Name':'Value'
+    })
+
+\ *Names* and *Values* must conform to the following rules:
+
+* Valid *Names* are utf-8 encoded strings with a maximum of 20 characters
+* Valid *Values* must be integers in the range of 0 through 2\ :sup:`32` - 1 (32-bit unsigned int)
```
Comment (nick-drozd) [inline docs/source/transaction_family_specifications/integerkey_transaction_family.rst]: Are there any rules on how an intkey TP should handle floats? More generally, should a TP reject a non-integer value outright, or can it convert it somehow?

File: docs/source/transaction_family_specifications/integerkey_transaction_family.rst (hunk 11 above)
Comment (dplumb94) [inline docs/source/transaction_family_specifications/integerkey_transaction_family.rst]: `[-64:]` is correct because we are taking the final 64 characters from the hash. I originally had an English description here but I took it out. I can add it back in if you think that would clarify matters.

Comment (nick-drozd) [inline docs/source/transaction_family_specifications/integerkey_transaction_family.rst]: I can never remember anything but the most basic Python string slicing syntax, so I think the English description should be included. I was under the impression that we were taking the **first** 64 characters of the hash, and I therefore thought that `[-64:]` was either a typo or a fancy trick that yielded the first 64 characters (but maybe that's just me?).

File: docs/source/transaction_family_specifications/integerkey_transaction_family.rst [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+.. code-block:: python
+
+    cbor.dumps({
+        // The name of the entry : The current value of the entry
+        'Name':'Value'
+    })
+
+\ *Names* and *Values* must conform to the following rules:
+
+* Valid *Names* are utf-8 encoded strings with a maximum of 20 characters
```
Comment (dplumb94) [inline docs/source/transaction_family_specifications/integerkey_transaction_family.rst]: This description contains the only restrictions for names.

File: docs/source/transaction_family_specifications/integerkey_transaction_family.rst (hunk 13 above)
Comment (dplumb94) [inline docs/source/transaction_family_specifications/integerkey_transaction_family.rst]: In the execution section, we mention that the value must be a 32-bit unsigned int or the transaction is invalid.

File: docs/source/transaction_family_specifications/integerkey_transaction_family.rst (hunk 11 above)
Comment (dplumb94) [inline docs/source/transaction_family_specifications/integerkey_transaction_family.rst]: I'm guessing that many folks have a similar recollection of Python string slicing. I'll add a concise English description.
